streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1435,6 +1435,34 @@ abstract class ZStream[-R, +E, +O](
   def groupedWithin(chunkSize: Long, within: Duration): ZStream[R with Clock, E, List[O]] =
     aggregateAsyncWithin(ZTransducer.collectAllN(chunkSize), Schedule.spaced(within))
 
+  /**
+   * Defines a time watermark for this stream. Removes all incoming values older then specified time delay.
+   * Watermarking is used to reduce memory and processing time for stale data.
+   *
+   * @param delayThreshold the delay to wait to data to arrive late
+   * @param timestamp time associated with the value, relative to the Unix epoch, in local timezone.
+   */
+  def watermark(delayThreshold: Duration, timestamp: O => java.time.Instant): ZStream[R with Clock, E, O] =
+    watermarkM(delayThreshold, a => ZIO.succeedNow(timestamp(a)))
+
+  /**
+   * Effectfull version of `watermark` function
+   *
+   * @param delayThreshold the delay to wait to data to arrive late
+   * @param timestamp time associated with the value, relative to the Unix epoch, in local timezone.
+   */
+  def watermarkM[R1 <: R, E1 >: E](
+    delayThreshold: Duration,
+    timestamp: O => ZIO[R1, E1, java.time.Instant]
+  ): ZStream[R1 with Clock, E1, O] =
+    filterM(event =>
+      for {
+        currentTime        <- clock.currentTime(ju.concurrent.TimeUnit.MILLISECONDS)
+        eventTimestamp     <- timestamp(event)
+        watermarkPredicate <- ZIO.succeedNow(eventTimestamp.toEpochMilli >= (currentTime - delayThreshold.toMillis))
+      } yield watermarkPredicate
```

Comment:
Looks like it could be simplified by  `yield (eventTimestamp.toEpochMilli >= (currentTime - delayThreshold.toMillis))`

Reply:
thanks, I fixed it